File: apps/frontend/src/mocks/lib/member-budget-quota.ts
import type { BudgetNode, Member, MemberBudgetQuota } from '@/api/types'
import { findBudgetNode, sumChildrenBudget } from '@/lib/budget'
import {
  getAllocatedKeyQuota,
  getPersonalQuota,
  getUsedKeyQuota,
  setPersonalQuota,
} from './member-quota'
import { findMemberById } from './query'

// Department member budget quota CRUD validation for MSW budget handlers.

export function getMemberQuotaCapacity(deptNode: BudgetNode): number {
  const reserved = deptNode.reservedPool ?? 0
  const childrenSum = sumChildrenBudget(deptNode)
  return Math.max(0, deptNode.budget - reserved - childrenSum)
}

export function buildMemberBudgetQuota(member: Member): MemberBudgetQuota {
  return {
    memberId: member.id,
    memberName: member.name,
    departmentId: member.departmentId,
    personalQuota: getPersonalQuota(member.id),
    allocated: getAllocatedKeyQuota(member.id),
    used: getUsedKeyQuota(member.id),
  }
}

export function validateMemberQuotaUpdate(
  budgetTree: BudgetNode[],
  members: Member[],
  memberId: string,
  personalQuota: number,
): string | null {
  const member = findMemberById(members, memberId)
  if (!member) return 'Member not found'

  const allocated = getAllocatedKeyQuota(memberId)
  if (personalQuota < allocated) {
    return `个人额度不能低于已分配 Key 额度（¥${allocated.toLocaleString()}）`
  }

  const deptNode = findBudgetNode(budgetTree, member.departmentId)
  if (!deptNode) return 'Department budget node not found'

  const capacity = getMemberQuotaCapacity(deptNode)
  const deptMembers = members.filter((m) => m.departmentId === member.departmentId)
  const otherSum = deptMembers
    .filter((m) => m.id !== memberId)
    .reduce((sum, m) => sum + getPersonalQuota(m.id), 0)

  if (otherSum + personalQuota > capacity) {
    return `超出部门可分配成员额度，当前剩余约 ¥${Math.max(0, capacity - otherSum).toLocaleString()}`
  }

  return null
}

export function applyMemberQuotaUpdate(
  members: Member[],
  memberId: string,
  personalQuota: number,
): MemberBudgetQuota {
  setPersonalQuota(memberId, personalQuota)
  const member = findMemberById(members, memberId)
  if (!member) {
    return {
      memberId,
      memberName: '',
      departmentId: '',
      personalQuota,
      allocated: getAllocatedKeyQuota(memberId),
      used: getUsedKeyQuota(memberId),
    }
  }
  return buildMemberBudgetQuota(member)
}

File: apps/backend/internal/pkg/budget/memberbudgetquota.go
package budget

import (
	"fmt"

	"github.com/tokenjoy/backend/internal/domain/types"
	"github.com/tokenjoy/backend/internal/pkg/org"
)

func GetMemberBudgetCapacity(deptNode types.BudgetNode) float64 {
	reserved := 0.0
	if deptNode.ReservedPool != nil {
		reserved = *deptNode.ReservedPool
	}
	childrenSum := SumChildrenBudget(deptNode)
	capacity := deptNode.Budget - reserved - childrenSum
	if capacity < 0 {
		return 0
	}
	return capacity
}

func BuildMemberBudgetQuota(member types.Member, platformKeys []types.PlatformKey) types.MemberBudgetQuota {
	return types.MemberBudgetQuota{
		MemberID: member.ID, MemberName: member.Name, DepartmentID: member.DepartmentID,
		PersonalBudget: member.PersonalBudget,
		Allocated:      GetAllocatedKeyBudget(platformKeys, member.ID),
		Used:           GetUsedKeyBudget(platformKeys, member.ID),
	}
}

func ValidateMemberBudgetUpdate(
	tree []types.BudgetNode,
	members []types.Member,
	platformKeys []types.PlatformKey,
	memberID string,
	personalBudget float64,
) *string {
	member, ok := org.FindMemberByID(members, memberID)
	if !ok {
		msg := "Member not found"
		return &msg
	}

	allocated := GetAllocatedKeyBudget(platformKeys, memberID)
	if personalBudget < allocated {
		msg := fmt.Sprintf("个人额度不能低于已分配 Key 额度（¥%s）", formatMoney(allocated))
		return &msg
	}

	deptNode := FindBudgetNode(tree, member.DepartmentID)
	if deptNode == nil {
		msg := "Department budget node not found"
		return &msg
	}

	capacity := GetMemberBudgetCapacity(*deptNode)
	otherSum := 0.0
	for _, m := range members {
		if m.DepartmentID == member.DepartmentID && m.ID != memberID {
			otherSum += GetPersonalBudget(members, m.ID)
		}
	}
	if otherSum+personalBudget > capacity {
		remaining := capacity - otherSum
		if remaining < 0 {
			remaining = 0
		}
		msg := fmt.Sprintf("超出部门可分配成员额度，当前剩余约 ¥%s", formatMoney(remaining))
		return &msg
	}
	return nil
}

func ApplyMemberBudgetUpdate(
	members []types.Member,
	platformKeys []types.PlatformKey,
	memberID string,
	personalBudget float64,
) (types.MemberBudgetQuota, []types.Member) {
	updatedMembers := SetMemberPersonalBudget(members, memberID, personalBudget)
	member, ok := org.FindMemberByID(updatedMembers, memberID)
	if !ok {
		return types.MemberBudgetQuota{
			MemberID: memberID, PersonalBudget: personalBudget,
			Allocated: GetAllocatedKeyBudget(platformKeys, memberID),
			Used:      GetUsedKeyBudget(platformKeys, memberID),
		}, updatedMembers
	}
	return BuildMemberBudgetQuota(*member, platformKeys), updatedMembers
}

func formatMoney(value float64) string {
	return fmt.Sprintf("%.0f", value)
}
